File: examples/backends/go/apierror.go
package main

import (
	"errors"
	"fmt"
	"net/http"

	meteroid "github.com/meteroid-oss/meteroid-clients/go"
)

// errorCode is the machine-readable identifier of the contract's `ErrorCode` schema.
type errorCode string

const (
	codeBadRequest   errorCode = "BAD_REQUEST"
	codeUnauthorized errorCode = "UNAUTHORIZED"
	codeNotFound     errorCode = "NOT_FOUND"
	// Reserved by the contract; no operation requires a subscription.
	codeNoSubscription          errorCode = "NO_SUBSCRIPTION"
	codeFeatureNotEntitled      errorCode = "FEATURE_NOT_ENTITLED"
	codeQuotaExhausted          errorCode = "QUOTA_EXHAUSTED"
	codeCheckoutUnavailable     errorCode = "CHECKOUT_UNAVAILABLE"
	codeCatalogNotSeeded        errorCode = "CATALOG_NOT_SEEDED"
	codeWebhookSignatureInvalid errorCode = "WEBHOOK_SIGNATURE_INVALID"
	codeUpstreamUnauthorized    errorCode = "UPSTREAM_UNAUTHORIZED"
	codeUpstreamError           errorCode = "UPSTREAM_ERROR"
	codeRateLimited             errorCode = "RATE_LIMITED"
	codeInternal                errorCode = "INTERNAL"
)

var statusByCode = map[errorCode]int{
	codeBadRequest:              http.StatusBadRequest,
	codeWebhookSignatureInvalid: http.StatusBadRequest,
	codeUnauthorized:            http.StatusUnauthorized,
	codeQuotaExhausted:          http.StatusPaymentRequired,
	codeFeatureNotEntitled:      http.StatusForbidden,
	codeNotFound:                http.StatusNotFound,
	codeNoSubscription:          http.StatusConflict,
	codeCheckoutUnavailable:     http.StatusConflict,
	codeRateLimited:             http.StatusTooManyRequests,
	codeInternal:                http.StatusInternalServerError,
	codeUpstreamUnauthorized:    http.StatusBadGateway,
	codeUpstreamError:           http.StatusBadGateway,
	codeCatalogNotSeeded:        http.StatusServiceUnavailable,
}

// apiError is the single error envelope of examples/openapi.yaml, and the only error
// type a handler returns on purpose.
//
// Quota and UpgradePlanCode carry no `omitempty`: both are always serialized, null
// where they do not apply, so a strict client never has to tell an absent key from a
// null one.
type apiError struct {
	Code            errorCode      `json:"code"`
	Message         string         `json:"message"`
	Quota           *QuotaSnapshot `json:"quota"`
	UpgradePlanCode *PlanCode      `json:"upgrade_plan_code"`

	// status is set only where the HTTP status is not the one the code implies (a 413).
	status int
}

func newAPIError(code errorCode, format string, args ...any) *apiError {
	return &apiError{Code: code, Message: fmt.Sprintf(format, args...)}
}

func badRequest(format string, args ...any) *apiError {
	return newAPIError(codeBadRequest, format, args...)
}

func unauthorized(message string) *apiError {
	return newAPIError(codeUnauthorized, "%s", message)
}

func catalogNotSeeded(format string, args ...any) *apiError {
	return newAPIError(codeCatalogNotSeeded, format, args...)
}

func internalError(format string, args ...any) *apiError {
	return newAPIError(codeInternal, format, args...)
}

// Error is the message alone: it is what the startup probe logs, and the code is
// already a field wherever an apiError is looked at as one.
func (e *apiError) Error() string { return e.Message }

func (e *apiError) httpStatus() int {
	if e.status != 0 {
		return e.status
	}
	return statusByCode[e.Code]
}

func (e *apiError) withQuota(quota QuotaSnapshot) *apiError {
	e.Quota = &quota
	return e
}

func (e *apiError) withUpgrade(plan *PlanCode) *apiError {
	e.UpgradePlanCode = plan
	return e
}

func (e *apiError) withStatus(status int) *apiError {
	e.status = status
	return e
}

// upstream translates an SDK failure into this contract's envelope. It sits in the
// `if err != nil` directly under the SDK call it belongs to:
//
//	customer, err := a.meteroid.Customers().GetCustomer(ctx, alias)
//	if err != nil {
//		return nil, upstream("GET /api/v1/customers/"+alias, err)
//	}
//
// The SDK returns a *meteroid.APIError for every non-2xx answer, and its StatusCode is
// what separates "your API key is wrong" (an operator problem) from "Meteroid is
// throttling" (retry) from everything else. When the body parsed as Meteroid's error
// envelope, Payload carries the typed code and message. Anything that is *not* an
// APIError never got an HTTP answer at all: a refused connection, a timeout, or a 2xx
// body the SDK could not decode.
func upstream(context string, err error) error {
	var already *apiError
	if errors.As(err, &already) {
		return already
	}

	var apiErr *meteroid.APIError
	if !errors.As(err, &apiErr) {
		return newAPIError(codeUpstreamError, "Could not reach Meteroid for %s: %v", context, err)
	}

	switch status := apiErr.StatusCode; status {
	case http.StatusUnauthorized, http.StatusForbidden:
		return newAPIError(codeUpstreamUnauthorized,
			"Meteroid rejected the API key on %s (HTTP %d). Check METEROID_API_KEY.", context, status)
	case http.StatusTooManyRequests:
		return newAPIError(codeRateLimited, "Meteroid responded 429 to %s. Retry shortly.", context)
	default:
		detail := string(apiErr.RawBody)
		switch {
		case apiErr.Payload != nil:
			detail = fmt.Sprintf("%s: %s", apiErr.Payload.Code, apiErr.Payload.Message)
		case apiErr.OAuthPayload != nil:
			description := ""
			if apiErr.OAuthPayload.ErrorDescription != nil {
				description = *apiErr.OAuthPayload.ErrorDescription
			}
			detail = fmt.Sprintf("%s: %s", apiErr.OAuthPayload.Error, description)
		}
		return newAPIError(codeUpstreamError, "Meteroid responded %d to %s: %s", status, context, truncate(detail))
	}
}

// isNotFound is true when the SDK error is an upstream 404 — "this object does not
// exist", which for a catalog lookup means "not seeded" rather than "Meteroid is broken".
func isNotFound(err error) bool {
	var apiErr *meteroid.APIError
	return errors.As(err, &apiErr) && apiErr.StatusCode == http.StatusNotFound
}

func truncate(body string) string {
	const limit = 300
	// Runes, not bytes: cutting a multi-byte character in half would not be JSON-safe.
	if runes := []rune(body); len(runes) > limit {
		return string(runes[:limit]) + "…"
	}
	return body
}

File: examples/backends/java/src/main/java/com/scribe/ApiError.java
package com.scribe;

import com.meteroid.exceptions.ApiException;

/**
 * The single error envelope of {@code examples/openapi.yaml} — the Java twin of the Rust backend's
 * {@code src/error.rs}.
 *
 * <p>Every non-2xx response in this backend is an {@code ApiError}. {@code quota} and {@code
 * upgrade_plan_code} are always serialized, {@code null} where they do not apply, so a strict
 * client never has to tell an absent key from a null one.
 *
 * <p>It is an unchecked exception so a handler can {@code throw} it from anywhere and one Javalin
 * exception mapper renders it — see {@link Main}.
 */
public final class ApiError extends RuntimeException {

    private static final long serialVersionUID = 1L;
    private static final int MAX_UPSTREAM_BODY = 300;

    private final ErrorCode code;
    private final String detail;
    private Dto.QuotaSnapshot quota;
    private PlanCode upgradePlanCode;

    public ApiError(ErrorCode code, String message) {
        super(message);
        this.code = code;
        this.detail = message;
    }

    public ApiError withQuota(Dto.QuotaSnapshot snapshot) {
        this.quota = snapshot;
        return this;
    }

    public ApiError withUpgrade(PlanCode plan) {
        this.upgradePlanCode = plan;
        return this;
    }

    // --- the serialized envelope -------------------------------------------------

    /**
     * What actually goes on the wire.
     *
     * <p>The exception itself is deliberately never handed to Jackson: {@link Throwable} carries
     * {@code cause}, {@code stackTrace} and {@code suppressed} getters, and the contract's error
     * schema is {@code additionalProperties: false}.
     *
     * @param code machine-readable identifier; clients switch on this, never on {@code message}
     * @param message human-readable explanation, safe to show in the demo UI
     * @param quota live quota snapshot, populated only for {@code QUOTA_EXHAUSTED}
     * @param upgradePlanCode the plan to offer as the fix, for {@code 402} and {@code 403}
     */
    public record Envelope(
            ErrorCode code, String message, Dto.QuotaSnapshot quota, PlanCode upgradePlanCode) {}

    public Envelope envelope() {
        return new Envelope(code, detail, quota, upgradePlanCode);
    }

    public ErrorCode code() {
        return code;
    }

    public int status() {
        return code.status();
    }

    // --- constructors for the common cases ---------------------------------------

    public static ApiError badRequest(String message) {
        return new ApiError(ErrorCode.BAD_REQUEST, message);
    }

    public static ApiError unauthorized(String message) {
        return new ApiError(ErrorCode.UNAUTHORIZED, message);
    }

    public static ApiError catalogNotSeeded(String message) {
        return new ApiError(ErrorCode.CATALOG_NOT_SEEDED, message);
    }

    public static ApiError internal(String message) {
        return new ApiError(ErrorCode.INTERNAL, message);
    }

    /**
     * Translate an SDK failure into this contract's envelope.
     *
     * <p>The SDK reports the upstream status on {@link ApiException#getCode()}, which is what
     * separates "your API key is wrong" (an operator problem) from "Meteroid is throttling" (retry)
     * from everything else.
     */
    public static ApiError upstream(String context, ApiException e) {
        int status = e.getCode();
        ErrorCode code;
        if (status == 401 || status == 403) {
            code = ErrorCode.UPSTREAM_UNAUTHORIZED;
        } else if (status == 429) {
            code = ErrorCode.RATE_LIMITED;
        } else {
            code = ErrorCode.UPSTREAM_ERROR;
        }

        String message;
        switch (code) {
            case UPSTREAM_UNAUTHORIZED:
                message =
                        "Meteroid rejected the API key on "
                                + context
                                + " (HTTP "
                                + status
                                + "). Check METEROID_API_KEY.";
                break;
            case RATE_LIMITED:
                message = "Meteroid responded 429 to " + context + ". Retry shortly.";
                break;
            default:
                message =
                        "Meteroid responded "
                                + status
                                + " to "
                                + context
                                + ": "
                                + truncate(e.getResponseBody());
                break;
        }
        return new ApiError(code, message);
    }

    /** The SDK throws {@link java.io.IOException} when it cannot reach Meteroid at all. */
    public static ApiError unreachable(String context, java.io.IOException e) {
        return new ApiError(
                ErrorCode.UPSTREAM_ERROR, "Could not reach Meteroid for " + context + ": " + e);
    }

    /**
     * True when the SDK error is an upstream 404 — "this object does not exist", which for a
     * catalog lookup means "not seeded" rather than "Meteroid is broken".
     */
    public static boolean isNotFound(ApiException e) {
        return e.getCode() == 404;
    }

    private static String truncate(String body) {
        if (body == null) {
            return "";
        }
        return body.length() <= MAX_UPSTREAM_BODY
                ? body
                : body.substring(0, MAX_UPSTREAM_BODY) + "…";
    }
}
